src/components/SignInButton.tsx

```diff
@@ -0,0 +1,212 @@
+    Metrics.captureEvent(eventList.userSignIn);
+  };
+
+  const handleRegistration = async (
+    redirectTo: string,
+    shouldRedirect: boolean
+  ) => {
+    try {
+      await registerAndRedirectNewUser(redirectTo, shouldRedirect);
+    } catch (error: unknown) {
+      await handleErrors(error as HttpError, redirectTo, shouldRedirect);
+    }
+  };
+
+  const registerAndRedirectNewUser = async (
+    redirectTo: string,
+    shouldRedirect: boolean
+  ) => {
+    const registeredUser = await User.registerUser();
+    setUserRoleStatuses(registeredUser, Storage);
+    Metrics.identify(Storage.getAnonymousId());
+    Metrics.syncProfile();
+    Metrics.captureEvent(eventList.userRegister);
+    history.push(
+      `/tos_acceptance${shouldRedirect ? `?redirectTo=${redirectTo}` : ''}`
+    );
+  };
+
+  const handleErrors = async (
+    error: HttpError,
+    redirectTo: string,
+    shouldRedirect: boolean
+  ) => {
+    const status = error.status;
+
+    switch (status) {
+      case 400:
+        setShowError(true);
+        setErrorInfo({
+          title: 'Error',
+          description: JSON.stringify(error),
+        });
+        break;
+      case 409:
+        handleConflictError(redirectTo, shouldRedirect);
+        break;
+      default:
+        setShowError(true);
+        setErrorInfo({
+          title: 'Error',
+          description: 'Unexpected error, please try again',
+        });
+        break;
+    }
+  };
+
+  const handleConflictError = async (
+    redirectTo: string,
+    shouldRedirect: boolean
+  ) => {
+    try {
+      await checkToSAndRedirect(shouldRedirect ? redirectTo : null);
+    } catch (error) {
+      Storage.clearStorage();
+    }
+  };
+
+  const onFailure = (error: Error) => {
+    Storage.clearStorage();
+    setIsLoading(false);
+    if (!error.message.includes('Popup closed by user')) {
+      setShowError(true);
+      setErrorInfo({
+        title: 'Error',
+        description: error.message,
+      });
+    }
+  };
+
+  const loadingIndicator = (): JSX.Element => {
+    return (
+      <span>
+        <img height='20px' src={loadingImage} />
+      </span>
+    );
+  };
+
+  const signInButton = (): JSX.Element => {
+    return (
+      <button
+        className={'btn-secondary'}
+        style={customStyle}
+        onClick={async () => {
+          setIsLoading(true);
+          await Auth.signIn(true, onSuccess, onFailure);
+          setIsLoading(false);
+        }}
+        disabled={isLoading}
+      >
+        {isLoading ? loadingIndicator() : 'Sign In'}
+      </button>
+    );
+  };
+
+  const errorAlert = (errorInfo: ErrorInfo): JSX.Element => {
+    return (
+      <div className='dialog-alert'>
+        <Alert
+          id='dialog'
+          type='danger'
+          title={errorInfo.title}
+          description={errorInfo.description}
+        />
+      </div>
+    );
+  };
+
+  return <div>{!showError ? signInButton() : errorAlert(errorInfo)}</div>;
```

Comment:
Nit: flipping the branches allows removing the `!`, reducing the amount of syntax a reader has to parse.

```suggestion
  return <div>{showError ? errorAlert(errorInfo) : signInButton()}</div>;
```